src/logic/basic.lean

```diff
@@ -1144,6 +1144,14 @@ hpq _ $ some_spec _
 noncomputable def subtype_of_exists {α : Type*} {P : α → Prop} (h : ∃ x, P x) : {x // P x} :=
 ⟨classical.some h, classical.some_spec h⟩
 
+/-- A version of `by_contradiction` that uses types instead of propositions. -/
+protected noncomputable def by_contradiction' {α : Sort*} (H : ¬ (α → false)) : α :=
+classical.choice $ peirce _ false $ λ h, (H $ λ a, h ⟨a⟩).elim
```

Comment:
Given `by_contradiction` is classical but not in the `classical` namespace, should this be in the root namespace too? I don't feel strongly.

Reply:
My thinking is to emphasize the relation to other equivalents of the axiom of choice (all of which are in the `classical` namespace). It's fine to be in a namespace since this is not likely to be a particularly commonly used theorem, it's mainly here as an "end goal" theorem, i.e. something which is proven just to demonstrate that it is true.